Implement ModuleParams inheritance in the MultiCopterLandDetector class.

mcsauder opened the pull request:
**Describe problem solved by the proposed pull request**
This PR implements `ModuleParams` inheritance in the `MultiCopterLandDetector` class in following #12209.

**Test data / coverage**
This PR was test flown on a 250 size quad with pixhawk 4 mini, all tests performed identical to current PX4:master branch land detect behaviors:
1) https://review.px4.io/plot_app?log=52aafa8a-aa07-4df7-8a3b-8ad48435718c
2) https://review.px4.io/plot_app?log=2c14b7cd-701a-4247-80f8-3fdbcffa1530
3) https://review.px4.io/plot_app?log=b571417e-0f10-4449-a4d9-88a6d71af96c

**Additional context**
This PR follows work in #12209 and streamlines remaining work for #9756.

@bkueng , @dagar . Please let me know if you have any questions on this PR.  Thank you!

-Mark

mcsauder commented:
@bkueng or @dagar, would you be willing to advise me on this `fmu-v2` error?
`error: 'MPC_LAND_SPEED' is not a member of 'px4::params'
`

dagar commented:
This is because `MPC_LAND_SPEED` comes from the MC position controller, which isn't included in the fixedwing only builds.

Short term the easy workaround is to keep using the old API for that particular param. One idea longer term would be to restructure this so that each vehicle type has a land detector with all the common code moving to a library.

mcsauder commented:
@dagar, would it be acceptable to move the param declarations/definitions required for the MulticopterLandDetector into 'land_detector_params_mc.c` ?  (I'm not thrilled about this solution, but it functions.)

See the latest commit I pushed up. Flight tested: https://review.px4.io/plot_app?log=e938ee50-cf2f-494c-909c-7b9461417686

mcsauder commented:
> Short term the easy workaround is to keep using the old API for that particular param.

I've implemented your suggestion, squashed commits, rebased against current upstream, and re-flight tested:
1) https://review.px4.io/plot_app?log=38dda5c6-dc15-48fa-baa1-73162a69de0b
2) https://review.px4.io/plot_app?log=06790127-8321-4daa-bb34-7af4c9c29e6b
3) https://review.px4.io/plot_app?log=98517189-bcc1-4d31-8e9c-f6a420eeaf08

dagar commented on src/modules/land_detector/MulticopterLandDetector.cpp:
```diff
@@ -103,25 +96,24 @@ void MulticopterLandDetector::_update_topics()
+	if (_param_update_sub.update(&param_update)) {
+		_update_params();
+		_freefall_hysteresis.set_hysteresis_time_from(false, (hrt_abstime)(1e6f * _param_lndmc_ffall_ttri.get()));
+		_param_lndmc_rot_max.set(math::radians(_param_lndmc_rot_max.get()));
```
I suppose you can get away with this, but it seems kind of shady (remote possibility of accidentally committing the value). I would do the math::radians call at runtime to keep it simple.

mcsauder replied:
I agree. Honestly, I am a little bit in the dark about why this is happening in PX4:master.  What can we do better here?  I have checked this value in particular because it feels like it is converting radians as though they are degrees to radians again.  This one left me scratching my head.  I'll gladly change it to anything appropriate.

mcsauder replied:
Let me study what happens if I delete that LOC and report back...

mcsauder replied:
Well, my easiest tests are straightforward; flight testing is fine and the value reports back what it should after flights and reboots:
logs:
https://review.px4.io/plot_app?log=79c21d7b-908f-43dc-9aaa-6c0be1de1898
https://review.px4.io/plot_app?log=3c1b9939-5720-4373-b8cb-a5ee03ed19de

![image](https://user-images.githubusercontent.com/2497951/60378519-d2cec100-99e0-11e9-9c37-25520600c071.png)

mcsauder replied:
You're right, that was wrong. Fixed, rebased against current master and squashed into one commit.

mcsauder commented on src/modules/land_detector/MulticopterLandDetector.cpp:
```diff
@@ -213,11 +204,11 @@ bool MulticopterLandDetector::_get_maybe_landed_state()
 
 	// Next look if all rotation angles are not moving.
-	float maxRotationScaled = _params.maxRotation_rad_s * landThresholdFactor;
+	float maxRotationScaled = math::radians(_param_lndmc_rot_max.get()) * landThresholdFactor;
```
@dagar, the issue you highlighted is addressed here.

mcsauder replied:
Flight test land detection works properly in the and the value checks out when queried: https://review.px4.io/plot_app?log=9c628e6e-b90e-4d1b-9308-e6b5a8cc15b5

mcsauder commented:
Rebased on today's master and flight tested again, detection works as it should and the same as current master:
 - https://review.px4.io/plot_app?log=74274cf1-acfc-49fe-a7cb-76149c078f1c
 - https://review.px4.io/plot_app?log=6f13a2b9-44ad-447b-ae32-f5885fd3168f

bkueng commented on src/modules/land_detector/MulticopterLandDetector.h:
```diff
@@ -97,35 +105,22 @@ class MulticopterLandDetector : public LandDetector
 
 	uORB::Subscription _actuator_controls_sub{ORB_ID(actuator_controls_0)};
 	uORB::Subscription _battery_sub{ORB_ID(battery_status)};
+	uORB::Subscription _param_update_sub{ORB_ID(parameter_update)};
```
Why do you need this? It is handled in the base class already.

mcsauder replied:
Thanks @bkueng!  Removed with upstream merge while resolving conflicts.

mcsauder commented:
Re-flight tested after latest commit: https://review.px4.io/plot_app?log=487a2f0c-d3c1-4c86-8fd4-8d23361b5811

bkueng commented on src/modules/land_detector/MulticopterLandDetector.cpp:
```diff
@@ -105,25 +98,23 @@ void MulticopterLandDetector::_update_topics()
-	_params.maxRotation_rad_s = math::radians(_params.maxRotation_rad_s);
+	parameter_update_s param_update;
+
+	if (_param_update_sub.update(&param_update)) {
```
You don't need to check for this, or call `_update_params()`. It is handled and called from the base class.
You can also keep `_param_update_sub` private.

mcsauder replied:
Great, thanks! That simplifies things. I pushed up that fix.

mcsauder commented:
Flight tested with latest commit and rebase against master: https://review.px4.io/plot_app?log=3b8762bc-156a-48a5-a5da-69c0425af7a4

mcsauder commented:
Thank you @bkueng for your review and time spent!  I appreciate it!
